proposals/20231220-features-adoption-and-deprecation.md

```diff
@@ -0,0 +1,207 @@
+
+- Syntax.
+- File format.
+- Schema (i.e., supported fields).
+- Elements that affect the way users can implement rules.
+- Elements that affect the way rules are triggered.
+
+However, any change related to the rule's output when triggered (i.e., the alert) is out of scope for this area (see next section).
+
+### Outputs/Alerts Area
+
+Alerts, delivered through Falco output channels, are Falco's primary output. The way and the format in which alerts are produced can have a significant impact on adopters. For example, removing a supported rule field also impacts this area, as adopters may have relied on that field when consuming Falco output.
+
+Any feature or behavior changes to the following aspects or elements are assumed to be user-facing changes to the Outputs/Alerts area:
+
+- Output and logging formats.
+- Schema of outputted data (i.e., supported fields).
+- Falco output channels.
+- Any element that might be consumed from the output.
+
+### Subsystem APIs (Plugins, gRPC Output, Metrics, etc.) Area
+
+Falco is also comprised of several subsystems providing specific APIs. These subsystems notably include plugin system API, gRPC output API, and metrics API.
+
+In the context of this proposal, only changes to **public APIs** are assumed to be user-facing changes to this area.
+
+Public APIs are defined as those supporting Falco functioning and explicitly intended for user usage. Internal APIs consumed by Falco or other tools are out of scope for this area. For instance, the driver APIs or libs APIs are intended to be mainly consumed by Falco and not by users.
+
+### Release Cycle
+
+In the context of this proposal, a release cycle is the period between two consecutive major or minor releases of Falco. Hotfix/Patch releases must not be counted.
+
+The actual duration of a release cycle can vary. Still, it's assumed to be about 16 weeks (as per our current defined [Release Cycles and Development Iterations](https://github.com/falcosecurity/falco/blob/master/proposals/20230511-roadmap-management.md#release-cycles-and-development-iterations)). In case of future modification to the Falco release schedule, a period of minimum 3 months must be assumed.
+
+## Proposal
+
+### Adoption Policy
+
+Feature additions will follow a straightforward adoption model with defined status levels (inspired by those we already have in place for [repositories](https://github.com/falcosecurity/evolution/blob/main/REPOSITORIES.md#status)) aiming to characterize the current maturity level.
+
+Each feature will have an assigned status level at any specific time (i.e., a Falco release). Statuses are shown in the table below.
+
+| Status | Feature Gate | Intended for |
+| --- | --- | --- |
+| Sandbox | Disabled by default | Experimental/alpha features, not recommended for production use, can be removed at any time without further notice. |
+| Incubating | Enabled by default | Beta features, long-term support is not guaranteed. |
```

Comment:
We should consider putting this off by default.

Reply:
Features like `metrics` are disabled by default because maybe not everyone needs or wants them, but should still be considered a "Stable" core feature. Another example is the different outputs channels. Only because we enabled one output channel by default, does not mean the other ones are not stable. 

I believe we need more nuances here. Allow stable features to be disabled by default. And Incubating features to be enabled by default. This will align with the rules maturity framework where the enabled tag is separate from the maturity level. Instead we can use terms like "shipped with the default package / configuration" or similar to let users know what options are enabled by default regardless of Status.

Reply:
Thanks to your comment, I've realized that what I wrote in the proposal is highly misleading :sweat_smile:  

I meant that we need to introduce a toggle to bulk turn all features on/off in a specific set (i.e., sandbox, incubating, etc.). This would be on top of a specific feature option that allows users to turn the specific feature on/off. The term "feature gate" probably is not accurate for the purpose I had in mind. Perhaps "maturity level toggle" works better. I will rework the text and fix that.

So, I totally agree that not all stable features must be enabled by default. I just meant that we don't need a "maturity level toggle" for stable features (because it wouldn't be helpful).

Reply:
ACK do you happen to know software that uses such bulk feature activation? Was thinking and unsure yet if I would find it useful for adoption (never heard of such a concept before). Usually you are used to toggle individual config option on and off.

Reply:
Removed the feature gate concept in the end. The policies should be simpler now.

https://github.com/falcosecurity/falco/pull/2986/commits/ef240a5568fb26d2b84ee7a90fbbcae8c88c5849